File: src/main/python/AoC2016_25.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 25
#

import sys

from aoc.assembunny import Assembunny
from aoc.assembunny import AssembunnyInstruction
from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.vm import Program
from aoc.vm import VirtualMachine

Input = list[AssembunnyInstruction]
Output1 = int
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, inputs: InputData) -> Input:
        return Assembunny.parse(tuple(inputs))

    def solve_vm(self, instructions: Input) -> int:
        vm_instructions = Assembunny.translate(instructions)

        def run_program(init_a: int) -> list[str]:
            output = list()
            program = Program(
                vm_instructions,
                inf_loop_treshold=6_000,
                output_consumer=lambda s: output.append(s),
            )
            program.set_register_value("a", init_a)
            try:
                VirtualMachine().run_program(program)
            finally:
                return output

        n = 150
        while True:
            output = run_program(n)
            if all(i % 2 == int(v) for i, v in enumerate(output)):
                return n
            n += 1

    def part_1(self, instructions: Input) -> int:
        return self.solve_vm(instructions)

    def part_2(self, instructions: Input) -> str:
        return "🎄"

    def samples(self) -> None:
        pass


solution = Solution(2016, 25)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2016_25.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.assembunny.Assembunny;
import com.github.pareronia.aoc.assembunny.Assembunny.AssembunnyInstruction;
import com.github.pareronia.aoc.solution.SolutionBase;
import com.github.pareronia.aoc.vm.Instruction;
import com.github.pareronia.aoc.vm.Program;
import com.github.pareronia.aoc.vm.VirtualMachine;
import com.github.pareronia.aoc.vm.VirtualMachine.InfiniteLoopException;

public final class AoC2016_25
        extends SolutionBase<List<AssembunnyInstruction>, Integer, String> {
    
    private AoC2016_25(final boolean debug) {
        super(debug);
    }
    
    public static AoC2016_25 create() {
        return new AoC2016_25(false);
    }
    
    public static AoC2016_25 createDebug() {
        return new AoC2016_25(true);
    }
    
    private List<Long> runProgram(
            final List<Instruction> vmInstructions,
            final Integer initA
    ) {
        final List<Long> output = new ArrayList<>();
        final Program program = new Program(
                vmInstructions,
                6_000,
                output::add);
        program.setRegisterValue("a", initA.longValue());
        try {
            new VirtualMachine().runProgram(program);
        } catch (final InfiniteLoopException e) {
            log(() -> program.getCycles());
            log(() -> output);
        }
        return output;
    }
    
    private Integer solveVM(final List<AssembunnyInstruction> instructions) {
        final List<Instruction> vmInstructions
                = Assembunny.translate(instructions);
        int n = 150;
        while(true) {
            log(n);
            final List<Long> output = runProgram(vmInstructions, n);
            if (Stream.iterate(0, i -> i < output.size(), i -> i + 1)
                        .allMatch(i -> i % 2 == output.get(i))) {
                return n;
            }
            n++;
        }
    }
    
    @Override
    protected List<AssembunnyInstruction> parseInput(final List<String> inputs) {
        return Assembunny.parse(inputs);
    }

    @Override
    public Integer solvePart1(final List<AssembunnyInstruction> instructions) {
        return solveVM(instructions);
    }
    
    @Override
    public String solvePart2(final List<AssembunnyInstruction> instructions) {
        return "🎄";
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2016_25.create().run();
    }
}